src/skin/qml/qmlcontrolproxy.cpp

```diff
@@ -0,0 +1,177 @@
+#include "skin/qml/qmlcontrolproxy.h"
+
+#include "moc_qmlcontrolproxy.cpp"
+
+namespace mixxx {
+namespace skin {
+namespace qml {
+
+namespace {
+constexpr double kDefaultValue = 0.0;
+constexpr double kDefaultParameter = 0.0;
+} // namespace
+
+QmlControlProxy::QmlControlProxy(QObject* parent)
+        : QObject(parent),
+          m_isComponentComplete(false),
+          m_pControlProxy(nullptr) {
+}
+
+void QmlControlProxy::componentComplete() {
+    m_isComponentComplete = true;
+    reinitializeOrReset();
+}
+
+bool QmlControlProxy::isKeyValid() const {
+    return m_coKey.isValid();
+}
+
+bool QmlControlProxy::isInitialized() const {
+    return m_pControlProxy != nullptr;
+}
+
+void QmlControlProxy::setGroup(const QString& group) {
+    if (m_coKey.group == group) {
+        return;
+    }
+
+    const bool keyValidBeforeSet = isKeyValid();
+    m_coKey.group = group;
+    emit groupChanged(group);
+
+    const bool keyValidAfterSet = isKeyValid();
+    if (keyValidBeforeSet != keyValidAfterSet) {
+        emit keyValidChanged(keyValidAfterSet);
+    }
+
+    reinitializeOrReset();
+}
+
+const QString& QmlControlProxy::getGroup() const {
+    return m_coKey.group;
+}
+
+void QmlControlProxy::setKey(const QString& key) {
+    if (m_coKey.item == key) {
+        return;
+    }
+
+    const bool keyValidBeforeSet = isKeyValid();
+    m_coKey.item = key;
+    emit keyChanged(key);
+
+    const bool keyValidAfterSet = isKeyValid();
+    if (keyValidBeforeSet != keyValidAfterSet) {
+        emit keyValidChanged(keyValidAfterSet);
+    }
+
+    reinitializeOrReset();
+}
+
+const QString& QmlControlProxy::getKey() const {
+    return m_coKey.item;
+}
+
+void QmlControlProxy::setValue(double newValue) {
+    if (!isInitialized()) {
+        emit valueChanged(kDefaultValue);
+        return;
+    }
+    m_pControlProxy->set(newValue);
+    slotControlProxyValueChanged(newValue);
+}
+
+double QmlControlProxy::getValue() const {
+    if (!isInitialized()) {
+        return kDefaultValue;
+    }
+    return m_pControlProxy->get();
+}
+
+void QmlControlProxy::setParameter(double newValue) {
+    if (!isInitialized()) {
+        emit parameterChanged(kDefaultValue);
+        return;
+    }
+    m_pControlProxy->setParameter(newValue);
+    emit valueChanged(m_pControlProxy->get());
+    emit parameterChanged(newValue);
+}
+
+double QmlControlProxy::getParameter() const {
+    if (!isInitialized()) {
+        return kDefaultParameter;
+    }
+    return m_pControlProxy->getParameter();
+}
+
+void QmlControlProxy::reinitializeOrReset() {
```

Comment:
Is orReset still valid? 
It look like the function returns early if it is already initialized

Reply:
OrReset is confusing, because resetting a control proxy is different from resetting a smart pointer. It is also not the alternative to initialize.

How about:
ReinitializeFromKey() 

Than it is clear that one cane assume that it  becomes invalid if the key is invalid

Reply:
Done.